error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/FileRules.java

```diff
@@ -63,4 +113,20 @@ File after(String prefix, String suffix) throws IOException {
       return Files.createTempFile(prefix, suffix).toFile();
     }
   }
+
+  /**
+   * Prefer {@link Files#createTempFile(Path,String, String, FileAttribute[])} over alternatives
+   * that create files with more liberal permissions.
+   */
+  static final class FilesCreateTempFileInCustomDirectoryToFile {
+    @BeforeTemplate
+    File before(File directory, String prefix, String suffix) throws IOException {
+      return File.createTempFile(prefix, suffix, directory);
```

Comment:
Ah, @timtebeek one thing to consider: `File.createTempFile(prefix, suffix, null)` is now matched by both rules. Not sure how OpenRewrite handles that? (Our custom Refaster integration resolves this kind of ambiguity by selecting the rule that produces the smallest replacement string, which in this case means that it (correctly) selects the rule above.)

Reply:
Interesting! I suspect the generated recipe will contain a MethodPattern matching exactly one of these, not either one.

Reply:
I suppose that if OpenRewrite applies them in the listed order, the first one will win. Anyway, generated code attached :smile:
<details>
<summary>Generated recipes</summary>
<code>
    /**
     * OpenRewrite recipe created for Refaster template {@code FileRules.FilesCreateTempFileToFile}.
     */
    @SuppressWarnings("all")
    @NonNullApi
    @Generated("org.openrewrite.java.template.processor.RefasterTemplateProcessor")
    public static class FilesCreateTempFileToFileRecipe extends Recipe {

        /**
         * Instantiates a new instance.
         */
        public FilesCreateTempFileToFileRecipe() {}

        @Override
        public String getDisplayName() {
            return "Prefer `Files#createTempFile(String, String, FileAttribute[])` over alternatives that create files with more liberal permissions";
        }

        @Override
        public String getDescription() {
            return "Recipe created for the following Refaster template:\n```java\nstatic final class FilesCreateTempFileToFile {\n    \n    @BeforeTemplate\n    @SuppressWarnings(value = {\"FilesCreateTempFileInCustomDirectoryToFile\", \"java:S5443\", \"key-to-resolve-AnnotationUseStyle-and-TrailingComment-check-conflict\"})\n    File before(String prefix, String suffix) throws IOException {\n        return Refaster.anyOf(File.createTempFile(prefix, suffix), File.createTempFile(prefix, suffix, null));\n    }\n    \n    @AfterTemplate\n    @SuppressWarnings(value = \"java:S5443\")\n    File after(String prefix, String suffix) throws IOException {\n        return Files.createTempFile(prefix, suffix).toFile();\n    }\n}\n```\n.";
        }

        @Override
        public TreeVisitor<?, ExecutionContext> getVisitor() {
            JavaVisitor<ExecutionContext> javaVisitor = new AbstractRefasterJavaVisitor() {
                final JavaTemplate before$0 = JavaTemplate
                        .builder("java.io.File.createTempFile(#{prefix:any(java.lang.String)}, #{suffix:any(java.lang.String)})")
                        .build();
                final JavaTemplate before$1 = JavaTemplate
                        .builder("java.io.File.createTempFile(#{prefix:any(java.lang.String)}, #{suffix:any(java.lang.String)}, null)")
                        .build();
                final JavaTemplate after = JavaTemplate
                        .builder("java.nio.file.Files.createTempFile(#{prefix:any(java.lang.String)}, #{suffix:any(java.lang.String)}).toFile()")
                        .build();

                @Override
                public J visitMethodInvocation(J.MethodInvocation elem, ExecutionContext ctx) {
                    JavaTemplate.Matcher matcher;
                    if ((matcher = before$0.matcher(getCursor())).find()) {
                        return embed(
                                after.apply(getCursor(), elem.getCoordinates().replace(), matcher.parameter(0), matcher.parameter(1)),
                                getCursor(),
                                ctx,
                                SHORTEN_NAMES
                        );
                    }
                    if ((matcher = before$1.matcher(getCursor())).find()) {
                        return embed(
                                after.apply(getCursor(), elem.getCoordinates().replace(), matcher.parameter(0), matcher.parameter(1)),
                                getCursor(),
                                ctx,
                                SHORTEN_NAMES
                        );
                    }
                    return super.visitMethodInvocation(elem, ctx);
                }

            };
            return Preconditions.check(
                    Preconditions.and(
                        new UsesType<>("java.io.File", true),
                        new UsesMethod<>("java.io.File createTempFile(..)")
                    ),
                    javaVisitor
            );
        }
    }

    /**
     * OpenRewrite recipe created for Refaster template {@code FileRules.FilesCreateTempFileInCustomDirectoryToFile}.
     */
    @SuppressWarnings("all")
    @NonNullApi
    @Generated("org.openrewrite.java.template.processor.RefasterTemplateProcessor")
    public static class FilesCreateTempFileInCustomDirectoryToFileRecipe extends Recipe {

        /**
         * Instantiates a new instance.
         */
        public FilesCreateTempFileInCustomDirectoryToFileRecipe() {}

        @Override
        public String getDisplayName() {
            return "Prefer `Files#createTempFile(Path,String, String, FileAttribute[])` over alternatives that create files with more liberal permissions";
        }

        @Override
        public String getDescription() {
            return "Recipe created for the following Refaster template:\n```java\nstatic final class FilesCreateTempFileInCustomDirectoryToFile {\n    \n    @BeforeTemplate\n    File before(File directory, String prefix, String suffix) throws IOException {\n        return File.createTempFile(prefix, suffix, directory);\n    }\n    \n    @AfterTemplate\n    File after(File directory, String prefix, String suffix) throws IOException {\n        return Files.createTempFile(directory.toPath(), prefix, suffix).toFile();\n    }\n}\n```\n.";
        }

        @Override
        public TreeVisitor<?, ExecutionContext> getVisitor() {
            JavaVisitor<ExecutionContext> javaVisitor = new AbstractRefasterJavaVisitor() {
                final JavaTemplate before = JavaTemplate
                        .builder("java.io.File.createTempFile(#{prefix:any(java.lang.String)}, #{suffix:any(java.lang.String)}, #{directory:any(java.io.File)})")
                        .build();
                final JavaTemplate after = JavaTemplate
                        .builder("java.nio.file.Files.createTempFile(#{directory:any(java.io.File)}.toPath(), #{prefix:any(java.lang.String)}, #{suffix:any(java.lang.String)}).toFile()")
                        .build();

                @Override
                public J visitMethodInvocation(J.MethodInvocation elem, ExecutionContext ctx) {
                    JavaTemplate.Matcher matcher;
                    if ((matcher = before.matcher(getCursor())).find()) {
                        return embed(
                                after.apply(getCursor(), elem.getCoordinates().replace(), matcher.parameter(2), matcher.parameter(0), matcher.parameter(1)),
                                getCursor(),
                                ctx,
                                SHORTEN_NAMES
                        );
                    }
                    return super.visitMethodInvocation(elem, ctx);
                }

            };
            return Preconditions.check(
                    Preconditions.and(
                        new UsesType<>("java.io.File", true),
                        new UsesMethod<>("java.io.File createTempFile(..)")
                    ),
                    javaVisitor
            );
        }
    }
</code>
</details>

Reply:
Figured a test would be quickest way to check, but my IDE is throwing up some error
> java: exporting a package from system module jdk.compiler is not allowed with --release

<details>
<summary>tech.picnic.errorprone.refasterrules.FileRulesRecipesTest</summary>

```java
package tech.picnic.errorprone.refasterrules;

import org.junit.jupiter.api.Test;
import org.openrewrite.test.RecipeSpec;
import org.openrewrite.test.RewriteTest;

import static org.openrewrite.java.Assertions.java;

final class FileRulesRecipesTest implements RewriteTest {
    @Override
    public void defaults(RecipeSpec spec) {
        spec.recipe(new FileRulesRecipes());
    }

    @Test
    void stringValueOf() {
        rewriteRun(
            //language=java
            java(
                """
                    import java.io.File;

                    class Test {
                        File twoArgs() throws IOException {
                            return File.createTempFile("foo", "bar");
                        }
                        File threeArgs() throws IOException {
                            return File.createTempFile("foo", "bar", new File("baz"));
                        }
                    }
                    """,
                """
                    import java.io.File;
                    import java.nio.file.Files;

                    class Test {
                        File twoArgs() throws IOException {
                            return Files.createTempFile("foo", "bar").toFile();
                        }
                        File threeArgs() throws IOException {
                            return Files.createTempFile(new File("baz").toPath(), "foo", "bar").toFile();
                        }
                    }
                    """
            )
        );
    }
}
```
</details>

Reply:
@timtebeek to fix that error; do the following: 

> If this happens, go to Settings -> Build, Execution, Deployment -> Compiler -> Java Compiler and deselect the option Use '--release' option for cross-compilation (Java 9 and later). See [IDEA-288052](https://youtrack.jetbrains.com/issue/IDEA-288052) for details.

It's listed in the README :).

Reply:
Thanks! With that the above test now passes. 🎉

Reply:
I was worried about the `File.createTempFile("foo", "bar", null)`, but when I modify the test like that it still passes :+1: